Specify browser by path

ldub opened the pull request:
Implements #98

I tested it with the following minimal config:

```
module.exports = {
  defaultBrowser: "com.google.Chrome",
  handlers: [
    { match: /apple/  , browser: "com.apple.Safari"         },
    { match: /mozilla/, browser: "/Applications/Firefox.app"}
  ]
};
```

As expected:
- https://apple.com opens in Safari
- https://google.com opens in Google
- https://mozilla.org opens in Firefox

I implemented the `bundleId` and `appPath` as `String?`. This makes the code a bit ugly in a few places as the compiler can't guarantee that one of the two is non-nil. Is there a better way to do this in Swift? Some kind of `Either` type, perhaps?

freshyill commented:
@ldub I've got a dummy question: Is does a "browser" have to be a web browser?

My use case: I'm getting sick of my browser filling up with tons of stupid Zoom pages and I want to try to pass them directly to Zoom.app. Any idea if that would work?

ldub commented:
@freshyill sure thing, I do this as well. 

This is possible with Finicky without needing my PR to be merged. You can specify any app by bundle id:

```
module.exports = {
  defaultBrowser: "com.google.Chrome",
  handlers: [
    { match: /zoom.us\/j\//, browser: "us.zoom.xos"}
  ]
};
```

If you need to find the bundle id of some app, just replace the app path here:

```
$ /usr/libexec/PlistBuddy -c 'Print CFBundleIdentifier' "/Applications/zoom.us.app/Contents/Info.plist"
us.zoom.xos
```

Note that links from within a browser will be opened in that same browser. However, a link from a native email client or slack will be routed by Finicky. If you'd like to route links inside a browser to Finicky you will need to install a chrome extension. https://github.com/dmarmor/epichrome can also help you do that.

johnste commented:
It would be interesting to look into creating browser extensions to rewrite urls to a finicky protocol to have them opened outside of the browser.

jdnicholas commented:
@ldub This may also be a potentially silly question/issue: 

I tried using cloning and building with your commits, but when using the following config, links opened when holding shift just open in the original Firefox app, instead of the duplicated "Work" browser as intended.

```
module.exports = {
  defaultBrowser: "Safari",
  handlers: [
    { match: ({ keys }) => keys.function, browser: "/Applications/Firefox.app"},
    { match: ({ keys }) => keys.shift, browser: "/Applications/WorkFirefox.app"},
  ]
};

```
(where WorkFirefox.app is where I've just dragged and renamed another copy of the Firefox app from the installer .dmg into my Applications folder)

Am I misunderstanding how this should potentially be working? Thanks!

ldub commented:
@jdnicholas - what happens when you run `open -a /Applications/WorkFirefox.app https://wikipedia.org` in the terminal?

ldub commented:
@johnste I had the same thought. I looked an found a minimal example of a browser extension that can communicate with a native app to redirect opening: https://github.com/kumar303/default-opener
Using a `finicky://` protocol like you suggest could actually be a little simpler.

Also, do you have any comments on this feature/PR itself?

jdnicholas commented:
> @jdnicholas - what happens when you run `open -a /Applications/WorkFirefox.app https://wikipedia.org` in the terminal?

@ldub That opens up Wikipedia in the WorkFirefox browser with no issues (and similarly `open -a /Applications/Firefox.app https://wikipedia.org` opens it in regular Firefox

ldub commented:
Hm, thanks for testing that out for me. I'll see if I can debug this a bit later today.

johnste commented:
I'll try to review the changes this weekend. Thanks!

jdnicholas commented:
Hi @ldub, thanks for checking. I went ahead and tried removing Finicky, then rebuilding your fork again, and this time, lo and behold, all working as intended! 

So, sorry about that, I guess there must have been something getting mixed up with my set up...

ldub commented:
@johnste ping, any chance you have some time soon to take a look?

johnste commented:
Sorry for the late reply @ldub, I had written the above review but forgot to post the draft.

ldub commented on Finicky/Finicky/AppDelegate.swift:
```diff
@@ -192,13 +192,17 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSUserNotificationCenterDele
+                        return browser
+                    }
                 }
+            } else if browser.appPath != nil {
```
Oh yeah, I thought I left a review comment about this but I guess not. I could not find any easy way to tell if an app is running by path. So that left me with two options:

if there are multiple matching options, one specified by bundle id and one by path, then:

1) always open one specified by path
2) never open the one specified by path

I chose #1 because the "specify by path" is an override in my mind. However, that may not be what you intended. What do you think it should do?

Also, I am not an advanced macOS app developer by any means. Do you know of a way to tell if an application is running given its path? Maybe some way to hook into activity monitor API? in the info of an app, the `Open Files and Ports` tab contains this info:

```
cwd
/
txt
/Applications/Google Chrome.app/Contents/MacOS/Google Chrome
```

Which, if I could read it, would help me determine if the app is running.

ldub replied:
I changed it to option 2 in my PR as that seems more in line with your intention.

ldub commented:
Hey @johnste I've adjusted the PR based on your feedback. I moved the appPath detection into the javascript api, where it considers anything starting with a `/` to be a path. I think thats a reasonable heuristic. 

I have opened a stackoverflow to ask if there is a way to pass `isAppDirectory` as a custom function into the JSContext: https://stackoverflow.com/questions/61315468/access-to-filesystem-in-a-mac-app-jscontext . It seems like maybe we could pass `isAppDirectory` through FinickyConfiApi like you do for logging? I'm not exactly sure how that works though so let me know if you think that could be done.

johnste commented:
From what I've seen so far it should definitely be possible. I will review your changes soon. Thanks!

johnste commented:
I think the merge request looks good. As for adding `isAppDirectory` I think it can wait until later, so I created a [new issue](https://github.com/johnste/finicky/issues/108) for that.

johnste commented:
Thank you so much @ldub, nice job!
